### ISSUE TITLE
CUB Presets Thread Local Compilation Error

### ISSUE THREAD

**Issue by masterav10** [bytedeco/javacpp-presets#1239]
I've got a presets library I'm working on for CUB [here](https://github.com/masterav10/immersed-presets). I updated my javacpp library to 1.5.8.SNAPHOT because I wanted to try out the new thread_local stuff in a different project. When compiling the JNI I get an unexpected error:

```
<...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.cu(201): error: thread-local variable cannot be dynamically initialized
```

This does not happen when I compile my decklink project. These are the differences:
> Decklink uses windows-x86_64.properties, and therefore CL.exe
> CUB uses windows-x86_64-cuda.properties, and therefore nvcc.exe

It looks to me like the command lines are nearly identical. I even tried including the windows preset just to make sure they had the same options available.

The full task I'm running can be found below:

```
> Task :immersed-presets-cub:javacppJniGenerate
Info: Generating <...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnijavacpp.cu
Info: Generating <...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.cu
Info: Compiling <...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.dll
Info: nvcc "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6\include" "-IC:\Program Files\Zulu\zulu8.40.0.25-ca-fx-jdk8.0.222-win_x64\include" "-IC:\Program Files\Zulu\zulu8.40.0.25-ca-fx-jdk8.0.222-win_x64\include\win32" "<...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.cu" "<...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnijavacpp.cu" -Xcompiler=/Oi,/O2,/EHsc,/Gy,/GL,/MD,/LD -std=c++11 --gpu-architecture=sm_61 -Xcompiler=-GL- -Xcompiler=/Zc:__cplusplus -Xptxas=-v -Xcompiler=/W3 -shared -o jnicub.dll "-LC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.6\lib\x64" cuda.lib cudart.lib psapi.lib
cl : Command line warning D9025 : overriding '/GL' with '/GL-'
jnicub.cu
<...>immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.cu(201): error: thread-local variable cannot be dynamically initialized

<...>\immersed-presets\immersed-presets-cub\build\javacpp\resources\windows-x86_64\jnicub.cu(270): warning #177-D: function "JavaCPP_createPointer" was declared but never referenced

1 error detected in the compilation of "<...>immersed-presets/immersed-presets-cub/build/javacpp/resources/windows-x86_64/jnicub.cu".
nvcc warning : The -std=c++11 flag is not supported with the configured host compiler. Flag will be ignored.
jnicub.cu
Exception in thread "main" java.lang.RuntimeException: Process exited with an error: 1
        at org.bytedeco.javacpp.tools.Builder.generateAndCompile(Builder.java:620)
        at org.bytedeco.javacpp.tools.Builder.build(Builder.java:1167)
        at org.bytedeco.javacpp.tools.Builder.main(Builder.java:1450)
```

Here is the block in question from jnicub.cu

```
193  #if !defined(NO_JNI_DETACH_THREAD) && defined(_WIN32)
194     static __declspec(thread) struct JavaCPP_thread_local {
195         JNIEnv* env = NULL;
196         ~JavaCPP_thread_local() {
197             if (env && JavaCPP_vm) {
198                 JavaCPP_vm->DetachCurrentThread();
199             }
200         }
201     } JavaCPP_thread_local; 
```

Any thoughts? Maybe there is a compilation flag I missed somewhere? I can copy this exact block into any of the CUDA samples that come with the toolkit and it works fine.

Standard Errata:

```
>nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2022 NVIDIA Corporation
Built on Tue_Mar__8_18:36:24_Pacific_Standard_Time_2022
Cuda compilation tools, release 11.6, V11.6.124
Build cuda_11.6.r11.6/compiler.31057947_0

>cl.exe
Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30145 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.
```

**Comment by saudet** [comment]
Sounds like some bug in MSVC. If it works fine when compiled elsewhere, it's probably caused by one of the flags that JavaCPP passes by default, but maybe shouldn't.

**Comment by masterav10** [comment]
This apparently works.

```
193  #if !defined(NO_JNI_DETACH_THREAD) && defined(_WIN32)
194     static struct JavaCPP_thread_local {
195         JNIEnv* env = NULL;
196         ~JavaCPP_thread_local() {
197             if (env && JavaCPP_vm) {
198                 JavaCPP_vm->DetachCurrentThread();
199             }
200         }
201     } thread_local JavaCPP_thread_local; 
```

I will make a pull request when I get home if you are fine with the change. It's probably some weird interaction between nvcc and cl. This change should make no difference in practice, since we are compiling with a minimum of c++14.

**Comment by saudet** [comment]
Sure, we can check for C++11 with `#if __cplusplus >= 201103L || _MSC_VER >= 1900"`, if you'd like to add that. In your case though, if you don't have any callback functions that you need to detach, you could also define `NO_JNI_DETACH_THREAD` to skip over that piece of code.

**Comment by masterav10** [comment]
So I did this:

```
@Properties(
        inherit = {cudart.class}, 
        names = {"windows-x86_64"}, 
        target = "org.bytedeco.cuda.cub", 
        global = "org.bytedeco.cuda.global.cub",
        value = {
            @Platform(
                include = {
                    ...
                },
                define = "NO_JNI_DETACH_THREAD"
            )
        }
)
@NoException
public class cub implements InfoMapper {
...
``` 
And it now fails when compiling jnijavacpp.cu. Is this supposed to happen? That file indeed does not have it defined.

**Comment by saudet** [comment]
How does it fail to compile? Any error messages?

**Comment by masterav10** [comment]
The exact same error message:

` error: thread-local variable cannot be dynamically initialized `

But it's coming from jnijavacpp.cu instead of jnicub.cu

**Comment by saudet** [comment]
Your fix has been released with JavaCPP 1.5.8. Thanks for the contribution!
